walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,115 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.ListIterator;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatizes return data structure and set duration parameter
+    //Collection<TimeRange> availTimes = new HashSet<>();
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+    final int length = (int) request.getDuration();
+    Collection<String>  requestAttendees = request.getAttendees();
+    
+    // Creates a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Collapse overlaping relevant events into one Time Range
+    ArrayList<TimeRange> collapsedBookTimes = collapseTimeConflicts(bookTimes);
+  
+    return getAvailableTimeRanges(collapsedBookTimes, length);
+  }
+
+/**Create an ArrayList of relevant events that will restrict available time ranges */
+  public ArrayList<TimeRange> findTimeConflicts(Collection<String> requestAttendees, Collection<Event> events){
+    // Set doesn't store duplicates and makes the for loop below more efficient
+    HashSet<TimeRange> bookTimes = new HashSet<TimeRange>();
+
+    // Iterates thorugh each event and decides if the TimeRange is relevant based on requestAttendees
+    for (Iterator<Event> eventIterator = events.iterator(); eventIterator.hasNext();){
+        Event event = eventIterator.next();
+        
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee)){
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+
+    // The times have to be returned as a list to be sorted by start time 
+    return new ArrayList(bookTimes);
+  }
+
+  public ArrayList<TimeRange> collapseTimeConflicts (ArrayList<TimeRange> bookTimes){
```

Comment:
This method both mutates its argument, and returns it. Consider making it either:

```
public List<TimeRange> collapseTimeConflicts(ImmutableList<TimeRange> bookTimes) {
  // ...
  return value;
}
```

or:

```
public void collapseTimeConflicts(List<TimeRange> bookTimes) {
  // mutate bookTimes directly
}
```

Reply:
So essentially, if I mutate the list I don't need to return it?